In [40]:
import pandas as pd
import numpy as np
import statistics
import matplotlib.pyplot as plt
from matplotlib.ticker import StrMethodFormatter

In [41]:
players= pd.read_csv(r"C:\Users\giada\Desktop\PROGETTO LAB\data2021\players_finale.csv")

In [42]:
players

player_id country_id                name hand     ht  byear_of_birth  \
0         105453        JPN       Kei Nishikori    R  178.0          1989.0   
1         106421        RUS     Daniil Medvedev    R  198.0          1996.0   
2         104542        FRA  Jo-Wilfried Tsonga    R  188.0          1985.0   
3         104871        FRA       Jeremy Chardy    R  188.0          1987.0   
4         200282        AUS      Alex De Minaur    R  183.0          1999.0   
...          ...        ...                 ...  ...    ...             ...   
10118     245144        ITA        Marco Miceli    U    NaN          1999.0   
10119     245145        GER      Marvin Moeller    U    NaN          2000.0   
10120     245146        GER   Justin Schlageter    U    NaN          2001.0   
10121     245147        BIH      Jasmina Tinjic    R    NaN          1991.0   
10122     245148        ARG        Tobias Sonne    U    NaN          2000.0   

       sex  
0        M  
1        M  
2      NaN  
3        M  
4        M  
...    ...  
10118    M  
10119    M  
10120    M  
10121    F  
10122    M  

[10123 rows x 7 columns]

In [43]:
#controlliamo se ci sono duplicati rispetto alla Primary key "player_id"
players.shape

(10123, 7)

In [44]:
players.drop_duplicates(subset ="player_id",
                     keep = False, inplace = True)

In [45]:
players.shape
#non ci sono duplicati

(10123, 7)

In [46]:
#vediamo quali sono i missing values
players.isnull().sum()

player_id            0
country_id           0
name                 0
hand                33
ht                9599
byear_of_birth    2110
sex                 34
dtype: int64

## Sex

In [47]:
#controlliamo chi sono le persone senza "sex"
players[players["sex"].isnull()]

player_id country_id                          name hand     ht  \
2        104542        FRA            Jo-Wilfried Tsonga    R  188.0   
31       106426        CHI                Cristian Garin    R  185.0   
53       104527        SUI                 Stan Wawrinka    R  183.0   
86       126156        FRA                 Antoine Hoang    R    NaN   
95       126203        USA                  Taylor Fritz    R  193.0   
104      111456        USA            Mackenzie McDonald    R    NaN   
110      106043        ARG             Diego Schwartzman    R  170.0   
136      126207        USA                Frances Tiafoe    R  188.0   
204      124079        ESP                Pedro Martinez    R    NaN   
230      144750        RSA                  Lloyd Harris    R  193.0   
319      105015        CRO                 Franko Skugor    R  193.0   
331      105223        ARG         Juan Martin del Potro    R  198.0   
371      200670        USA                     J.J. Wolf    R    NaN   
399      123755        COL            Daniel Elahi Galan    R  191.0   
532      110778        BRA                  Joao Menezes    R    NaN   
568      127300        UZB               Jurabek Karimov    U    NaN   
577      122669        PER           Juan Pablo Varillas    R    NaN   
587      208029        DEN                   Holger Rune    R    NaN   
714      126939        UZB              Khumoun Sultanov    R    NaN   
776      126185        RUS               Evgenii Tiurnev    R    NaN   
1272     206403        UKR                  Alona Fomina    R    NaN   
2027     216569        TUR        Zeynep  Sena Sarioglan    U    NaN   
2180     106331        AUS         Christopher O'Connell    R  183.0   
2187     105032        AUS                     Sam Groth    R    NaN   
2733     103926        ESP       Daniel Munoz de la Nava    L    NaN   
2775     122298        NED       Botic van de Zandschulp    R    NaN   
2972     105586        BUL               Alexandar Lazov    U    NaN   
3114     110474        IRL                  David O'Hare    U    NaN   
3352     127326        ISR                    Ben Patael    U    NaN   
3884     106238        ESP  Andres Artunedo Martinavarro    R    NaN   
7270     209404        ESP            Marc Othman Ktiri     U    NaN   
7714     200707        USA                 Max Colorado     R  183.0   
8155     209890        CRO                Mili Poljicak     U    NaN   
8162     209984        SVK                  Milos Karol     R    NaN   

      byear_of_birth  sex  
2             1985.0  NaN  
31            1996.0  NaN  
53            1985.0  NaN  
86            1995.0  NaN  
95            1998.0  NaN  
104           1996.0  NaN  
110           1993.0  NaN  
136           1999.0  NaN  
204           1998.0  NaN  
230           1998.0  NaN  
319           1988.0  NaN  
331           1989.0  NaN  
371           1999.0  NaN  
399           1997.0  NaN  
532           1997.0  NaN  
568           1998.0  NaN  
577           1996.0  NaN  
587           2003.0  NaN  
714           1999.0  NaN  
776           1997.0  NaN  
1272          1989.0  NaN  
2027          2002.0  NaN  
2180          1995.0  NaN  
2187          1988.0  NaN  
2733          1982.0  NaN  
2775          1996.0  NaN  
2972          1990.0  NaN  
3114          1991.0  NaN  
3352          1998.0  NaN  
3884          1994.0  NaN  
7270          2004.0  NaN  
7714          1996.0  NaN  
8155             NaN  NaN  
8162          2003.0  NaN

É stato possibile controllare sul sito peoplepill (https://peoplepill.com/people/alona-fomina) come tutte queste persone senza Sex siamo "male", tranne per "Alona". 
Si procederà quindi a sostituire "nan" con il corrispettivo sesso

In [48]:
#rimpiazziamo tutti con "M"
players["sex"].fillna("M", inplace = True)

In [49]:
display(players.iloc[1272][6])

'M'

In [50]:
#poi sostituiamo solo "Alona"
players.at[1272, 'sex'] = 'F'

In [51]:
#controlliamo "Alona"
display(players.iloc[1272][6])

'F'

In [52]:
#controlliamo "Andres Artunedo Martinavarro"
display(players.iloc[3884][6])

'M'

## HAND

Poichè in hand troviamo un valore "U" che probabilmente significherà "undefined", sostituiamo tutte le u con "Nan", per poi sostituirli in base alla distribuzione di R e L sul totale di partecipanti 

In [53]:
players["hand"].value_counts()

U    6286
R    3402
L     402
Name: hand, dtype: int64

In [54]:
#sostituire i valori U con "NAN"
players['hand'] = players['hand'].replace("U",np.nan)

In [55]:
players.isnull().sum()

player_id            0
country_id           0
name                 0
hand              6319
ht                9599
byear_of_birth    2110
sex                  0
dtype: int64

In [56]:
#vedo la distribuzione dell'hand
players["hand"].value_counts()

R    3402
L     402
Name: hand, dtype: int64

In [57]:
#vediamo la percentuale di distribuzione sul totale di hand 
players['hand'].value_counts()/players['hand'].value_counts().sum()

R    0.894322
L    0.105678
Name: hand, dtype: float64

Poichè la distribuzione di hand è all'incirca 90% right e 10% left,  verrà sostituita l'hand, randomicamente, per il totale dei record secondo queste distribuzioni.

In [58]:
from random import sample
from random import seed

In [59]:
seed(2)

In [60]:
#ottengo gli indici delle righe in cui hand è missing
indici = list(players[players['hand'].isna()].index)

In [61]:
#prendo il 90% random degli indici da sostituire con 'R'
right = sample(indici, int(0.9*len(indici)))

In [62]:
#Metto R in base all'indice
for i in right:
    players.loc[i,'hand'] = 'R'

In [63]:
#Gli ultimi missing values che rimangono sono il 10% in cui mettere 'L'
players['hand'] = players['hand'].replace(np.nan, 'L')

In [64]:
#controlliamo
players.isnull().sum()

player_id            0
country_id           0
name                 0
hand                 0
ht                9599
byear_of_birth    2110
sex                  0
dtype: int64

##  HT

Raggruppiamo l'altezza in base al CONTINENT e sex, successivamente calcoliamo la media, per poi rimpiazzare i valori mancanti

In [65]:
#importiamo geography
geo= pd.read_csv(r"C:\Users\giada\Desktop\PROGETTO LAB\data2021\geography.csv")

In [66]:
#creiamo un dataframe chiamato "merged" che continene tutte le colonne di players e continent
merged= pd.merge(players, geo, how='inner', left_on = 'country_id', right_on = 'country_code')
merged.drop(columns=['language','country_code'], inplace=True)

In [67]:
merged

player_id country_id                 name hand     ht  byear_of_birth  \
0         105453        JPN        Kei Nishikori    R  178.0          1989.0   
1         106034        JPN    Yasutaka Uchiyama    R    NaN          1992.0   
2         211768        JPN          Naomi Osaka    R  180.0          1997.0   
3         106121        JPN          Taro Daniel    R    NaN          1993.0   
4         106415        JPN   Yoshihito Nishioka    L  170.0          1996.0   
...          ...        ...                  ...  ...    ...             ...   
10118     239179        LVA  Elizabete Klavinska    R    NaN             NaN   
10119     223102        LVA        Nicole Geller    R    NaN             NaN   
10120     223104        LVA        Alise Vindiga    R    NaN             NaN   
10121     238211        CRI   Elena Salazar Roca    R    NaN             NaN   
10122     223434        BAN        Jhilik Chakma    R    NaN             NaN   

      sex continent  
0       M      Asia  
1       M      Asia  
2       F      Asia  
3       M      Asia  
4       M      Asia  
...    ..       ...  
10118   F    Europe  
10119   F    Europe  
10120   F    Europe  
10121   F   America  
10122   F      Asia  

[10123 rows x 8 columns]

In [68]:
players['ht'] = merged.groupby(['sex','continent'])['ht'].transform(lambda x: x.fillna(x.mean()))

In [69]:
#controlliamo
players.isnull().sum()

player_id            0
country_id           0
name                 0
hand                 0
ht                   0
byear_of_birth    2110
sex                  0
dtype: int64

## Controlliamo se ci sono valori strani nell'altezza

In [70]:
#controlliamo le altezze per vedere se ci sono valori "strani", cioè possibili outliers o errori di scrittura
column = players["ht"]

max_value = column.max()
min_value= column.min()

print(max_value)
print(min_value)

211.0
2.0


In [71]:
#vediamo se ci sono persone più basse di 150 cm
players.loc[players['ht'] < 150]

player_id country_id         name hand     ht  byear_of_birth sex
500      201664        TUR  Pemra Ozgen    R    2.0          1986.0   F
2161     104330        USA   Rajeev Ram    R  145.0          1984.0   M

É stato possibile notare come sia presente una persona con altezza 145 ed una con altezza 2. Probabilmente quest'ultimo caso è dovuto ad un errore di battitura.
Per tal ragione, è stato cercato il nome della tennista per sostituire la sua altezza con quella reale (fonte:https://it.wikipedia.org/wiki/Pemra_%C3%96zgen). Inoltre è stata anche verificata l'altezza del secondo player, dal medesimo sito, per poi correggerla. 

In [72]:
players.loc[players['ht'] == 2.0, 'ht'] = 170.0
players.loc[players['name'] == "Rajeev Ram", 'ht'] = 193.0

In [73]:
players.loc[players['ht'] < 150]

Empty DataFrame
Columns: [player_id, country_id, name, hand, ht, byear_of_birth, sex]
Index: []

In [74]:
#controlliamo
players.isnull().sum()

player_id            0
country_id           0
name                 0
hand                 0
ht                   0
byear_of_birth    2110
sex                  0
dtype: int64

In [75]:
#ricontrollo se non ci sono duplicati
players.shape

(10123, 7)

## BYEAR OF BIRTH

Poichè non è possibile risalire alle altre date di nascita tramite data integration, allora, è stato deciso di lasciare le righe che non presentano "byearofbirth" poichè un qualsiasi metodo di sostituzione dei dati porterebbe solo ad una distorsione eccessiva e inoltre, qualora avessimo rimosso le righe ci sarebbe stata un inconsistenza nella chiave esterna in match. Di conseguenza, i nan values sono stati sostituiti con "-1".

In [76]:
players['byear_of_birth'] = players['byear_of_birth'].replace(np.nan, -1)

## Convertiamo anno di nascita e ht in "int"

In [77]:
players['byear_of_birth'] = players['byear_of_birth'].astype(int)
players['ht'] = players['ht'].astype(int)

In [78]:
#salvare il dataset, come "player_clean"
players.to_csv(r'C:\Users\giada\Desktop\PROGETTO LAB\tab da inserire\players_clean.csv', index = False, header = True)